blogs/contractionorder.md

```diff
@@ -0,0 +1,338 @@
+
+![](/assets/treewidth_figs/treedecomposition_2.png)
+
+where the left graph is the original graph and the right one is the tree decomposition of the graph, and the tree width is 2.
+
+
+To find the optimal tree decomposition of a simple graph, one can use the Bouchitté-Todinca algorithm [^Bouchitté], which is discussed in detail in the blog [Finding the Optimal Tree Decomposition with Minimal Treewidth](/blogs/treewidth/), one can refer to it for more details.
+
+### From Tree Decomposition to Contraction Order
+
+Then how tree decomposition and tree width are related to the contraction order of the tensor network?
+In this sub-section, we will answer the following questions: how to get a contraction order from a tree decomposition? And why the contraction order is optimal if the tree width is minimal?
+
+For the first question, according to the definition of tree decomposition, we can obtain vertices elimination order by traversing the tree decomposition in a bottom-up manner. 
+For each node of the tree decomposition, a vertex can be eliminated if it is in the bag of the node and not in the bags of its parent node, as shown in the figure below:
+
+![](/assets/treewidth_figs/elimination_order2.png)
+
+and the order is $\{\{i, j, l\}, \{k\}, \{m\}, \{n\}\}$, where the last vertex to be eliminated first.
+
+The elimination order of the indices can then be used to determine the contraction order of the tensor network, where two tensors are contracted if they share a common index and the index is eliminated.
+According to the elimination order above, the contraction is shown below:
+
+![](/assets/treewidth_figs/elimination.png)
+
+Of course, the contraction order is not unique by selecting different node as the root of the decomposition tree, different contraction orders can be obtained.
+
+Then to answer the second question, we have to introduce a property of the optimal tree decomposition:
+
+**Proposition 1**: all tree bags of the optimal tree decomposition are cliquish, i.e., the vertices in the same bag are either connected or belong to the same separator.
+
+If two indices are connected in the line graph, then they at least belong to a same tensor.
+For the separators, notice that in a contraction order from a tree decomposition, a separator is exactly the indices of an intermediate tensor, so that they should also be treated as connected.
+In the example above, $BE$ has indices $ik$, corresponding to the separator $\{i, k\}$ between tree bags $\{i, j, k\}$ and $\{i, k, m\}$, and so on.
+
+Therefore, in a tree bag all indices are "connected", in each step of the contraction, we will have to loop over all the indices in the same bag, so that the bottleneck of time complexity is exactly by $O(e^{tw(G) + 1})$, and the since all intermediate tensors are characterized by the separators, and separators are real subset of tree bags, the space complexity bounded by $O(e^{tw(G)})$.
```

Comment:
I can understand the time complexity, but for the space complexity:

Is it clear that the separators = intersection of two tree bags always has the width as size, i.e. that the intersection has one index less then the individual bags (since the width is one less then the size). Does this mean that any two connected bags in the tree decomposition differ by just a single index, as in the example case, or is that an artefact of this specific example?

Reply:
This should be an artifact of this specific example, in general cases the size of the separators can be smaller that the width, so that the width is only an upper bound of the size of the separators.

Here is an example:
```Julia
julia> using TreeWidthSolver, Graphs

julia> g = smallgraph(:petersen)
{10, 15} undirected simple Int64 graph

julia> td = decomposition_tree(g)
tree width: 4.0
tree decomposition:
Set([5, 6, 7, 3, 1])
├─ Set([7, 2, 3, 1])
├─ Set([5, 4, 6, 7, 3])
│  └─ Set([4, 6, 7, 9])
└─ Set([5, 6, 7, 10, 3])
   └─ Set([6, 10, 8, 3])
```
where the intersection of `Set([5, 6, 7, 10, 3])` and `Set([6, 10, 8, 3])` is `Set([3, 6, 10])`, which is smaller that $4$.